haystack/nodes/other/shaper.py

```diff
@@ -0,0 +1,419 @@
+        invalid_vars = [key for key in self.inputs.keys() if key not in invocation_context]
+        if invalid_vars:
+            raise ValueError(
+                f"The following variables, specified in Shaper directives, were not resolved {invalid_vars}."
+                f"Please check the inputs definition and try again."
+            )
+
+        # traverse the inputs YAML subtree of the Shaper component declaration, invoke the specified
+        # functions, and store the results in the invocation context
+        self.traverse(None, self.inputs, invocation_context, self.invoke)  # type: ignore
+        invocation_context.pop("invocation_stack")
+
+        meta = meta or {}
+        meta["invocation_context"] = invocation_context
+        response = {
+            "query": invocation_context.pop("query", query),
+            "documents": invocation_context.pop("documents", documents),
+            "file_paths": invocation_context.pop("file_paths", file_paths),
+            "labels": invocation_context.pop("labels", labels),
+            "meta": {**meta},
+        }, "output_1"
+        return response
+
+    def run_batch(
+        self,
+        queries: Optional[Union[str, List[str]]] = None,
+        file_paths: Optional[List[str]] = None,
+        labels: Optional[Union[MultiLabel, List[MultiLabel]]] = None,
+        documents: Optional[Union[List[Document], List[List[Document]]]] = None,
+        meta: Optional[Union[Dict[str, Any], List[Dict[str, Any]]]] = None,
+        params: Optional[dict] = None,
+        debug: Optional[bool] = None,
+    ):
+        multi_query = isinstance(queries, list) and len(queries) > 0 and isinstance(queries[0], str)
+        single_query = isinstance(queries, str)
+        multi_docs_list = isinstance(documents, list) and len(documents) > 0 and isinstance(documents[0], list)
+        single_docs_list = isinstance(documents, list) and len(documents) > 0 and isinstance(documents[0], Document)
+
+        results = {"queries": queries, "documents": documents, "meta": meta}
+        if single_query:
+            if single_docs_list:
+                response, _ = self.run(query=queries, documents=documents, meta=meta)  # type: ignore
+                self.result_update_helper(results, response)
+                results.update(response)
+            elif multi_docs_list:
+                for doc in documents:  # type: ignore
+                    response, _ = self.run(query=queries, documents=doc, meta=meta)  # type: ignore
+                    self.result_update_helper(results, response)
+        elif multi_query:
+            for query, docs in zip(queries, documents):  # type:ignore
+                response, _ = self.run(query=query, documents=docs, meta=meta)  # type: ignore
+                self.result_update_helper(results, response)
+        return results, "output_1"
+
+    @staticmethod
+    def result_update_helper(parent_dict, response):
```

Comment:
If you use a `defaultdict(list)` for `response`, you can get rid of most (if not all) the logic in this method: `parent_dict[k].append(v)` would never fail.